crypto/o_str.c

```diff
@@ -90,6 +90,69 @@ size_t OPENSSL_strlcat(char *dst, const char *src, size_t size)
     return l + OPENSSL_strlcpy(dst, src, size);
 }
 
+/**
+ * @brief Converts a string to an unsigned long integer.
+ *
+ * This function attempts to convert a string representation of a number
+ * to an unsigned long integer, given a specified base. It also provides
+ * error checking and reports whether the conversion was successful.
+ * This function is just a wrapper around the POSIX strtoul function with
+ * additional error checking.  This implies that errno for the caller is set
+ * on calls to this function.
+ *
+ * @param str The string containing the representation of the number.
+ * @param endptr A pointer to a pointer to character. If not NULL, it is set
+ *               to the character immediately following the number in the
+ *               string.
+ * @param base The base to use for the conversion, which must be between 2,
+ *             and 36 inclusive, or be the special value 0. If the base is 0,
+ *             the actual base is determined by the format of the initial
+ *             characters of the string.
+ * @param num A pointer to an unsigned long where the result of the
+ *            conversion is stored.
+ *
+ * @return 1 if the conversion was successful, 0 otherwise. Conversion is
+ *         considered unsuccessful if no digits were consumed or if an error
+ *         occurred during conversion.
+ *
+ * @note It is the caller's responsibility to check if the conversion is
+ *       correct based on the expected consumption of the string as reported
+ *       by endptr.
+ */
+int OPENSSL_strtoul(const char *str, char **endptr, int base,
+                    unsigned long *num)
+{
+    char *tmp_endptr;
+    char **internal_endptr = endptr == NULL ? &tmp_endptr : endptr;
+
+    errno = 0;
+
+    *internal_endptr = (char *)str;
+
+    if (num == NULL)
+        return 0;
+
+    if (str == NULL)
+        return 0;
+
+    /* Fail on negative input */
+    if (*str == '-')
+        return 0;
+
+    *num = strtoul(str, internal_endptr, base);
+    /*
+     * If no digits were consumed in the translation
+     * or if there was a reported error during the translation
+     * fail the operation, otherwise succeed.
+     * Note, it's the responsibility of the caller to determine if the
+     * conversion was correct based on the expected consumption of the string
+     * as reported by endptr
+     */
+    if (str == (const char *)*internal_endptr || errno != 0)
```

Comment:
```suggestion
    if (str == *internal_endptr || errno != 0)
```
The casting is not required here. The `==` operator checks whether the two pointers point to the same memory location, not whether the strings they point to are equal.

Reply:
gcc was giving me warnings here about const discardment.  I grant thats a false positive, since its a comparison, rather than an assignment, but not doing so breaks --strict-warnings, so I think its better to do the cast here, as its a no-op and quiets a false warning

Reply:
It seems a bit off. Could you please double-check it once more?

Reply:
hmm, yeah, I must have been thinking of the other cast above, this is fine.

Reply:
Nevermind - your comment was added while I was testing